### ISSUE TITLE
prefer-expect-assertions - allow `beforeEach(expect.hasAssertions)`

### ISSUE THREAD

**Issue by EvgenyOrekhov** [jest-community/eslint-plugin-jest#876]
Recently I found out that instead of repeating `expect.hasAssertions()` in each test, you can have a single `beforeEach(expect.hasAssertions)` call at the top of the file. 

It would be great if `prefer-expect-assertions` could detect file-level `beforeEach(expect.hasAssertions)` call and NOT give warnings in this case.

Current behavior:

```js
beforeEach(expect.hasAssertions);

test("foo", () => {
  // Warning: Every test should have either `expect.assertions(<number of assertions>)`
  // or `expect.hasAssertions()` as its first expression
});
```

```js
beforeEach(() => {
  expect.hasAssertions();
  someOtherFunction();
});

test("foo", () => {
  // Warning: Every test should have either `expect.assertions(<number of assertions>)`
  // or `expect.hasAssertions()` as its first expression
});
```

Expected behavior:

```js
beforeEach(expect.hasAssertions);

test("foo", () => {
  // No warnings
});
```

```js
beforeEach(() => {
  expect.hasAssertions();
  someOtherFunction();
});

test("foo", () => {
  // No warnings
});
```

**Comment by G-Rath** [comment]
I can do you one better: you can stick that [`beforeEach` in a script](https://github.com/G-Rath/audit-app/blob/main/test/setupExpectEachTestHasAssertions.ts) to be called by `jest` via [`setupFilesAfterEnv`](https://github.com/G-Rath/audit-app/blob/main/jest.config.ts#L15) - I do it in most of my projects when I'm just using `expect`.

------

Overall, I think I'm going to say "maybe" to this - I actually think we should get very close to being able to determine this, but have concerns over edge-cases and the complexity of the code it'd require.

Specifically, because we will have to track the nesting of each `beforeEach` relative to the test functions to cover i.e

```
describe('MyClass', () => {
  describe('#myMethod', () => {
    beforeEach(expect.hasAssertions);

    it('does something', () => {
      // ...
    });
  });

  describe('#myOtherMethod', () => {
    it('does something', () => {
      // ...
    });
  });
});
```

Still I think it might be possible based on some of the logic we have in other rules that do similar tracking of nesting that we might be able to do it.
